akka-http-tests/src/test/scala/akka/http/scaladsl/server/directives/RouteDirectivesSpec.scala

```diff
@@ -157,6 +158,30 @@ class RouteDirectivesSpec extends AnyFreeSpec with GenericRoutingSpec {
     }
   }
 
+  "the fromFunction directive" should {
+    "use a function to complete a request" in {
+      Get(Uri("https://akka.io/foo")) ~> {
+        fromFunction(req => Future.successful(HttpResponse(OK, entity = req.uri.toString)))
+      } ~> check { response shouldEqual HttpResponse(200, entity = "https://akka.io/foo") }
+    }
+    "reject the request when the future fails" in {
+      Get(Uri("https://akka.io/foo")) ~> {
+        concat(
+          fromFunction(req => Future.failed(new IllegalStateException("Some error"))),
+          complete(ImATeapot)
+        )
+      } ~> check { response shouldEqual HttpResponse(500, entity = "There was an internal server error.") }
+    }
+    "reject the request when the function throws fails" in {
+      Get(Uri("https://akka.io/foo")) ~> {
+        concat(
+          fromFunction(req => throw new IllegalStateException("Some error")),
+          complete(ImATeapot)
+        )
+      } ~> check { response shouldEqual HttpResponse(500, entity = "There was an internal server error.") }
+    }
+  }
+
```

Comment:
Maybe we could also support the `RejectionError` to reject from such a function? Though, seems a bit weird mix of APIs if you go that way...